File: wasm/statistics/src/sensitivity.rs
/// Sensitivity analysis — simulate parameter changes and their impact on profit.
///
/// # Mathematical model
/// - Sales = Customers × Transaction Value
/// - Gross Profit = Gross Sales - Cost - Discount - Cost Inclusion
/// - Each delta is applied independently (customer × txValue has multiplicative interaction)
///
/// # Invariants
/// - Zero deltas → simulatedGrossProfit == baseGrossProfit
/// - grossProfitDelta == simulatedGrossProfit - baseGrossProfit
/// - salesDelta == simulatedSales - totalSales

fn safe_divide(num: f64, den: f64, fallback: f64) -> f64 {
    if den != 0.0 { num / den } else { fallback }
}

pub struct SensitivityBase {
    pub total_sales: f64,
    pub total_cost: f64,
    pub total_discount: f64,
    pub gross_sales: f64,
    pub total_customers: f64,
    pub total_cost_inclusion: f64,
    pub budget: f64,
    pub elapsed_days: f64,
    pub sales_days: f64,
}

pub struct SensitivityDeltas {
    pub discount_rate_delta: f64,
    pub customers_delta: f64,
    pub transaction_value_delta: f64,
    pub cost_rate_delta: f64,
}

pub struct SensitivityResult {
    pub base_gross_profit: f64,
    pub base_gross_profit_rate: f64,
    pub simulated_gross_profit: f64,
    pub simulated_gross_profit_rate: f64,
    pub gross_profit_delta: f64,
    pub simulated_sales: f64,
    pub sales_delta: f64,
    pub simulated_projected_sales: f64,
    pub projected_sales_delta: f64,
    pub budget_achievement_delta: f64,
}

pub fn calculate_sensitivity(base: &SensitivityBase, deltas: &SensitivityDeltas) -> SensitivityResult {
    let base_tx_value = safe_divide(base.total_sales, base.total_customers, 0.0);
    let base_cost_rate = safe_divide(base.total_cost, base.gross_sales, 0.0);
    let base_discount_rate = safe_divide(base.total_discount, base.gross_sales, 0.0);

    let sim_customers = base.total_customers * (1.0 + deltas.customers_delta);
    let sim_tx_value = base_tx_value * (1.0 + deltas.transaction_value_delta);
    let sim_sales = sim_customers * sim_tx_value;

    let sim_discount_rate = base_discount_rate + deltas.discount_rate_delta;
    let sim_gross_sales = safe_divide(sim_sales, 1.0 - sim_discount_rate, sim_sales);

    let sim_cost_rate = base_cost_rate + deltas.cost_rate_delta;
    let sim_cost = sim_gross_sales * sim_cost_rate;
    let sim_discount = sim_gross_sales * sim_discount_rate.max(0.0);

    let cost_inclusion_rate = safe_divide(base.total_cost_inclusion, base.total_sales, 0.0);
    let sim_consumable = sim_sales * cost_inclusion_rate;

    let base_gp = base.gross_sales - base.total_cost - base.total_discount - base.total_cost_inclusion;
    let sim_gp = sim_gross_sales - sim_cost - sim_discount - sim_consumable;

    let base_gp_rate = safe_divide(base_gp, base.total_sales, 0.0);
    let sim_gp_rate = safe_divide(sim_gp, sim_sales, 0.0);

    let daily_avg_base = safe_divide(base.total_sales, base.elapsed_days, 0.0);
    let daily_avg_sim = safe_divide(sim_sales, base.elapsed_days, 0.0);
    let base_projected = daily_avg_base * base.sales_days;
    let sim_projected = daily_avg_sim * base.sales_days;

    let base_achievement = safe_divide(base_projected, base.budget, 0.0);
    let sim_achievement = safe_divide(sim_projected, base.budget, 0.0);

    SensitivityResult {
        base_gross_profit: base_gp,
        base_gross_profit_rate: base_gp_rate,
        simulated_gross_profit: sim_gp,
        simulated_gross_profit_rate: sim_gp_rate,
        gross_profit_delta: sim_gp - base_gp,
        simulated_sales: sim_sales,
        sales_delta: sim_sales - base.total_sales,
        simulated_projected_sales: sim_projected,
        projected_sales_delta: sim_projected - base_projected,
        budget_achievement_delta: sim_achievement - base_achievement,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_base() -> SensitivityBase {
        SensitivityBase {
            total_sales: 10_000_000.0,
            total_cost: 7_500_000.0,
            total_discount: 200_000.0,
            gross_sales: 10_200_000.0,
            total_customers: 5000.0,
            total_cost_inclusion: 50_000.0,
            budget: 12_000_000.0,
            elapsed_days: 15.0,
            sales_days: 30.0,
        }
    }

    fn zero_deltas() -> SensitivityDeltas {
        SensitivityDeltas {
            discount_rate_delta: 0.0,
            customers_delta: 0.0,
            transaction_value_delta: 0.0,
            cost_rate_delta: 0.0,
        }
    }

    #[test]
    fn invariant_zero_deltas_no_change() {
        let base = make_base();
        let result = calculate_sensitivity(&base, &zero_deltas());
        assert!(
            (result.gross_profit_delta).abs() < 1e-6,
            "Zero deltas should produce zero GP delta, got {}",
            result.gross_profit_delta
        );
        assert!(
            (result.sales_delta).abs() < 1e-6,
            "Zero deltas should produce zero sales delta"
        );
    }

    #[test]
    fn invariant_delta_identity() {
        let base = make_base();
        let deltas = SensitivityDeltas {
            discount_rate_delta: -0.01,
            customers_delta: 0.05,
            transaction_value_delta: 0.0,
            cost_rate_delta: 0.0,
        };
        let result = calculate_sensitivity(&base, &deltas);
        let expected_delta = result.simulated_gross_profit - result.base_gross_profit;
        assert!(
            (result.gross_profit_delta - expected_delta).abs() < 1e-6,
            "grossProfitDelta must equal simulated - base"
        );
    }

    #[test]
    fn customer_increase_raises_sales() {
        let base = make_base();
        let deltas = SensitivityDeltas {
            customers_delta: 0.10, // +10%
            ..zero_deltas()
        };
        let result = calculate_sensitivity(&base, &deltas);
        assert!(result.sales_delta > 0.0, "10% more customers should increase sales");
    }

    #[test]
    fn cost_rate_improvement_raises_profit() {
        let base = make_base();
        let deltas = SensitivityDeltas {
            cost_rate_delta: -0.01, // 1pt improvement
            ..zero_deltas()
        };
        let result = calculate_sensitivity(&base, &deltas);
        assert!(result.gross_profit_delta > 0.0, "Lower cost rate should improve GP");
    }
}

File: app/src/domain/calculations/algorithms/sensitivity.ts
/**
 * 感度分析（Sensitivity Analysis）
 *
 * StoreResult の主要指標に対して、入力パラメータの変動が
 * 粗利・売上にどの程度影響するかをシミュレーションする純粋関数群。
 *
 * 弾性値（Elasticity）: パラメータ1pt変動あたりの粗利変動額
 *
 * @responsibility R:unclassified
 */
import { z } from 'zod'
import { safeDivide } from '../utils'

// ─── Zod Schemas ─────────────────────────────────────

/** 感度分析の入力パラメータ（ベースライン値） */
export const SensitivityBaseSchema = z.object({
  totalSales: z.number(),
  totalCost: z.number(),
  totalDiscount: z.number(),
  grossSales: z.number(),
  totalCustomers: z.number(),
  totalCostInclusion: z.number(),
  averageMarkupRate: z.number(),
  budget: z.number(),
  elapsedDays: z.number(),
  salesDays: z.number(),
})
export type SensitivityBase = z.infer<typeof SensitivityBaseSchema>

/** 変動量（差分） — 各パラメータの変動pt/% */
export const SensitivityDeltasSchema = z.object({
  discountRateDelta: z.number(),
  customersDelta: z.number(),
  transactionValueDelta: z.number(),
  costRateDelta: z.number(),
})
export type SensitivityDeltas = z.infer<typeof SensitivityDeltasSchema>

/** 感度分析の結果 */
export const SensitivityResultSchema = z.object({
  baseGrossProfit: z.number(),
  baseGrossProfitRate: z.number(),
  simulatedGrossProfit: z.number(),
  simulatedGrossProfitRate: z.number(),
  grossProfitDelta: z.number(),
  simulatedSales: z.number(),
  salesDelta: z.number(),
  simulatedProjectedSales: z.number(),
  projectedSalesDelta: z.number(),
  budgetAchievementDelta: z.number(),
})
export type SensitivityResult = z.infer<typeof SensitivityResultSchema>

/** 弾性値（各パラメータ1pt変動あたりの粗利変動額） */
export const ElasticityResultSchema = z.object({
  discountRateElasticity: z.number(),
  customersElasticity: z.number(),
  transactionValueElasticity: z.number(),
  costRateElasticity: z.number(),
})
export type ElasticityResult = z.infer<typeof ElasticityResultSchema>

// ─── Functions ────────────────────────────────────────

/**
 * 感度分析を実行する。
 *
 * ベースライン値に対してdeltasで指定された変動を適用し、
 * 粗利・売上の変化をシミュレーションする。
 *
 * 計算ロジック:
 * - 売上 = 客数 × 客単価
 * - 粗利 = 売上 - 原価 - 売変 - 消耗品
 * - 各deltaは独立に適用（相互作用項あり: 客数×客単価は乗算効果）
 * @calc-id CALC-022
 */
export function calculateSensitivity(
  base: SensitivityBase,
  deltas: SensitivityDeltas,
): SensitivityResult {
  const baseTxValue = safeDivide(base.totalSales, base.totalCustomers, 0)
  const baseCostRate = safeDivide(base.totalCost, base.grossSales, 0)
  const baseDiscountRate = safeDivide(base.totalDiscount, base.grossSales, 0)

  // シミュレーション後の値
  const simCustomers = base.totalCustomers * (1 + deltas.customersDelta)
  const simTxValue = baseTxValue * (1 + deltas.transactionValueDelta)
  const simSales = simCustomers * simTxValue

  // 粗売上は売上 + 売変 の関係から推定（売変率の変動を反映）
  const simDiscountRate = baseDiscountRate + deltas.discountRateDelta
  const simGrossSales = safeDivide(simSales, 1 - simDiscountRate, simSales)

  // 原価 = 粗売上 × 原価率
  const simCostRate = baseCostRate + deltas.costRateDelta
  const simCost = simGrossSales * simCostRate

  // 売変額
  const simDiscount = simGrossSales * Math.max(0, simDiscountRate)

  // 消耗品は売上比例で推定
  const costInclusionRate = safeDivide(base.totalCostInclusion, base.totalSales, 0)
  const simConsumable = simSales * costInclusionRate

  // 粗利 = 売上 - 原価 - 消耗品
  // 推定法ベース: margin = coreSales - estMethodCogs
  // 簡略化: grossProfit ≈ grossSales - cost - discount - costInclusion
  const baseGrossProfit =
    base.grossSales - base.totalCost - base.totalDiscount - base.totalCostInclusion
  const simGrossProfit = simGrossSales - simCost - simDiscount - simConsumable

  const baseGPRate = safeDivide(baseGrossProfit, base.totalSales, 0)
  const simGPRate = safeDivide(simGrossProfit, simSales, 0)

  // 着地予測
  const dailyAvgBase = safeDivide(base.totalSales, base.elapsedDays, 0)
  const dailyAvgSim = safeDivide(simSales, base.elapsedDays, 0)
  const baseProjected = dailyAvgBase * base.salesDays
  const simProjected = dailyAvgSim * base.salesDays

  // 予算達成率の変化
  const baseAchievement = safeDivide(baseProjected, base.budget, 0)
  const simAchievement = safeDivide(simProjected, base.budget, 0)

  return {
    baseGrossProfit,
    baseGrossProfitRate: baseGPRate,
    simulatedGrossProfit: simGrossProfit,
    simulatedGrossProfitRate: simGPRate,
    grossProfitDelta: simGrossProfit - baseGrossProfit,
    simulatedSales: simSales,
    salesDelta: simSales - base.totalSales,
    simulatedProjectedSales: simProjected,
    projectedSalesDelta: simProjected - baseProjected,
    budgetAchievementDelta: simAchievement - baseAchievement,
  }
}

/**
 * 弾性値を計算する。
 *
 * 各パラメータを1単位（1pt or 1%）変動させたときの粗利変動額を算出。
 * 実務的な「このパラメータにテコ入れすべきか」の判断材料。
 */
export function calculateElasticity(base: SensitivityBase): ElasticityResult {
  const onePt = 0.01

  const discountResult = calculateSensitivity(base, {
    discountRateDelta: -onePt, // 売変率1pt改善
    customersDelta: 0,
    transactionValueDelta: 0,
    costRateDelta: 0,
  })

  const customersResult = calculateSensitivity(base, {
    discountRateDelta: 0,
    customersDelta: onePt, // 客数1%増
    transactionValueDelta: 0,
    costRateDelta: 0,
  })

  const txValueResult = calculateSensitivity(base, {
    discountRateDelta: 0,
    customersDelta: 0,
    transactionValueDelta: onePt, // 客単価1%増
    costRateDelta: 0,
  })

  const costResult = calculateSensitivity(base, {
    discountRateDelta: 0,
    customersDelta: 0,
    transactionValueDelta: 0,
    costRateDelta: -onePt, // 原価率1pt改善
  })

  return {
    discountRateElasticity: discountResult.grossProfitDelta,
    customersElasticity: customersResult.grossProfitDelta,
    transactionValueElasticity: txValueResult.grossProfitDelta,
    costRateElasticity: costResult.grossProfitDelta,
  }
}

/**
 * StoreResult から SensitivityBase を抽出する。
 * Application層で使用するヘルパー。domain層の型変換のみ。
 */
export function extractSensitivityBase(result: {
  readonly totalSales: number
  readonly totalCost: number
  readonly totalDiscount: number
  readonly grossSales: number
  readonly totalCustomers: number
  readonly totalCostInclusion: number
  readonly averageMarkupRate: number
  readonly budget: number
  readonly elapsedDays: number
  readonly salesDays: number
}): SensitivityBase {
  return {
    totalSales: result.totalSales,
    totalCost: result.totalCost,
    totalDiscount: result.totalDiscount,
    grossSales: result.grossSales,
    totalCustomers: result.totalCustomers,
    totalCostInclusion: result.totalCostInclusion,
    averageMarkupRate: result.averageMarkupRate,
    budget: result.budget,
    elapsedDays: result.elapsedDays,
    salesDays: result.salesDays,
  }
}
